src/arfsdao.ts

```diff
@@ -611,58 +671,121 @@ export class ArFSDAO extends ArFSDAOAnonymous {
+
+		while (hasNextPage) {
+			const gqlQuery = buildQuery(
+				[
+					{ name: 'Drive-Id', value: driveId },
+					{ name: 'Entity-Type', value: 'folder' }
+				],
+				cursor
+			);
+
+			const response = await this.arweave.api.post(graphQLURL, gqlQuery);
+			const { data } = response.data;
+			const { transactions } = data;
+			const { edges } = transactions;
+			hasNextPage = transactions.pageInfo.hasNextPage;
+
+			const folders: Promise<ArFSPrivateFolder>[] = edges.map(async (edge: GQLEdgeInterface) => {
+				cursor = edge.cursor;
+				const { node } = edge;
+				const folderBuilder = await ArFSPrivateFolderBuilder.fromArweaveNode(
+					node,
+					this.arweave,
+					this.wallet as JWKWallet,
+					drivePassword
+				);
+				return await folderBuilder.build(node);
 			});
+			allFolders.push(...(await Promise.all(folders)));
+		}
 
-			// Get the drives transaction ID
-			drive.txId = node.id;
+		return allFolders;
+	}
+
+	async getPrivateFilesWithParentFolderIds(folderIDs: FolderID[], drivePassword: string): Promise<ArFSPrivateFile[]> {
+		let cursor = '';
+		let hasNextPage = true;
+		const allFiles: ArFSPrivateFile[] = [];
+
+		while (hasNextPage) {
+			const gqlQuery = buildQuery(
+				[
+					{ name: 'Parent-Folder-Id', value: folderIDs },
+					{ name: 'Entity-Type', value: 'file' }
+				],
+				cursor
+			);
+
+			const response = await this.arweave.api.post(graphQLURL, gqlQuery);
+			const { data } = response.data;
+			const { transactions } = data;
+			const { edges } = transactions;
+			hasNextPage = transactions.pageInfo.hasNextPage;
+			const files: Promise<ArFSPrivateFile>[] = edges.map(async (edge: GQLEdgeInterface) => {
+				const { node } = edge;
+				cursor = edge.cursor;
+				const fileBuilder = await ArFSPrivateFileBuilder.fromArweaveNode(
+					node,
+					this.arweave,
+					this.wallet as JWKWallet,
+					drivePassword
+				);
+				return await fileBuilder.build(node);
+			});
+			allFiles.push(...(await Promise.all(files)));
 		}
-		return drive.build(this.wallet as JWKWallet, drivePassword, this.arweave);
+		return allFiles;
+	}
+
+	/**
+	 * Lists the children and self of certain private folder
+	 * @param {FolderID} folderId the folder ID to list children of
+	 * @returns {ArFSPrivateFileOrFolderWithPaths[]} an array representation of the children and parent folder
+	 */
+	async listPrivateFolder(folderId: FolderID, password: string): Promise<ArFSPrivateFileOrFolderWithPaths[]> {
+		const folder = await this.getPrivateFolder(folderId, password);
+
+		// Fetch all of the folder entities within the drive
+		const driveIdOfFolder = folder.driveId;
+		const allFolderEntitiesOfDrive = (await this.getAllFoldersOfPrivateDrive(driveIdOfFolder, password)).filter(
+			lastRevisionFilter
+		);
+
+		// Feed entities to FolderHierarchy.setupNodesWithEntity()
+		const hierarchy = FolderHierarchy.newFromEntities(allFolderEntitiesOfDrive);
+		const folderIDs = hierarchy.allFolderIDs();
```

Comment:
Why is this allFolderIds rather than the subtree of folders starting from folderId?

Reply:
If we found a way of working with the entities no matter their privacy, we would be able to update the code in a single place, reducing the rate of this type of mistake :thinking: 

I gotta push the same changes I made tho the public method

Reply:
Totally agree. My gears are turning on how to do that. But for now let's get things working before we jump to making them elegant.

Reply:
161c1d2e0757f91c81ba74835a495509a3e16fa7 updates it